text/0249-v2-experiments.md

```diff
@@ -0,0 +1,380 @@
+import * as cognito_preview from 'aws-cdk-lib-experiments/aws-cognito';
+
+const idp = new cognito_preview.UserPoolIdentityProviderOidc(this, 'OIDC', { ... });
+const supported = [cognito.UserPoolClientIdentityProvider.custom("MyProviderName")];
+const userPoolClient = new cognito.UserPoolClient(...);
+```
+
+Advantages:
+
+- Very explicit (customer has to add a dependency on a package with
+  "experiments" in the name and version `0.x`).
+- It's possible for unstable modules to depend on other unstable modules.
+
+Disadvantages:
+
+- It's not possible for stable modules to depend on unstable ones. This has
+  serious side implications:
+  - All unstable modules that have stable dependents today will have to be
+    graduated before `v2.0` is released.
+  - Before a module is graduated, all of its dependencies need to be graduated.
+  - It will not be possible to add new dependencies on unstable modules to
+    stable modules in the future (for example, that's a common need for
+    StepFunction Tasks).
+- Graduating a module to stable will be a breaking change for customers. We can
+  mitigate this downside by keeping the old unstable module around, but that
+  leads to duplicated classes.
+
+### Option 3: separate multiple unstable packages
+
+In this option, each experimental library will be vended as a separate package.
+Each would have the name "experiments" in it (possible naming convention:
+`@aws-cdk-lib-experiments/aws-<service>`), and would be released in version
+`0.x` to make it absolutely obvious this is unstable code. Each package would
+declare a caret peer dependency on `aws-cdk-lib`.
+
+So, the above Cognito example would be:
+
+```ts
+import * as cognito from 'aws-cdk-lib/aws-cognito';
+import * as cognito_preview from '@aws-cdk-lib-experiments/aws-cognito';
+
+const idp = new cognito_preview.UserPoolIdentityProviderOidc(this, 'OIDC', { ... });
+const supported = [cognito.UserPoolClientIdentityProvider.custom("MyProviderName")];
+const userPoolClient = new cognito.UserPoolClient(...);
+```
+
+Advantages:
+
+- Very explicit (customer has to add a dependency on a package with
+  "experiments" in the name and version `0.x`).
+- This is closest to the third-party CDK package experience our customers will
+  have.
+
+Disadvantages:
+
+- It's not possible for stable modules to depend on unstable ones (with the same
+  implications as above).
+- It's not possible for unstable modules to depend on other unstable modules, as
+  doing that brings us back to the dependency hell that mono-CDK was designed to
+  solve.
+- Graduating a module to stable will be a breaking change for customers. We can
+  mitigate this downside by keeping the old unstable package around, but that
+  leads to duplicated classes.
+
+### Option 4: separate V3 that's all unstable
+
+In this option, we will fork the CDK codebase and maintain 2 long-lived
+branches: one for version `2.x`, which will be all stable, and one for version
```

Comment:
Not a fan of this option. Maintaining 2 long-lived branches is not easy and is going to cause us a lot of maintainance pain. Separate repos or folders are much easier to manage.

Reply:
Plus, the disadvantages section points the other problems quite well.